### PR TITLE
chore: test walker

### PR DESCRIPTION
**Pull request by aladdin-add**
todo: 1. fix types(any) 2. better error message

### CODE REVIEW

**Comment by KFlash** [comment]
The CST specs are complete now, so no mayor changes will be done except for the types, so do not try to "walk" the type cst nodes yet.

**Comment by KFlash** [comment]
@aladdin-add Any progress? Be aware that the `NodeFlag` change the behavior of the parser. You may get some surprises, specially when it comes to the type system CST nodes ;)

**Comment by KFlash** [comment]
@aladdin-add Friendly ping. Any progress?

**Comment by KFlash** [comment]
@aladdin-add Friendly ping. Any progress?

**Comment by KFlash** [comment]
@aladdin-add check the readme! I exported all CST nodes so it can be tested in the walker.

**Comment by aladdin-add** [comment]
sounds great!

sorry, I've been very busy on some rl stuff, but will try to get it asap.

**Comment by KFlash** [comment]
It's okay. I may be offline the next 15 days myself due to some rl stuff and not available.
Anyway. I think the parser API at least is okay now.

**Comment by aladdin-add** [comment]
yes! I found it's very usefule. seems some nodes are missing `.end`, can you fix it?

**Comment by KFlash** [comment]
@aladdin-add It's impossible that the nodes are missing '.end'. How can that be?  Every CST node is created with a start and end value

File: src/forEachChild.ts [hunk 1]
```diff
@@ -107,8 +107,8 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
     case SyntaxKind.ArrowFunction:
       return (
         visitNode(node.arrowToken, node, visitor) ||
-        visitNode(node.typeParameters, node, visitor) ||
-        visitNode(node.parameters, node, visitor) ||
+        visitNodes(node.typeParameters, node, visitor) ||
```
Comment (KFlash) [inline src/forEachChild.ts:110]: should not 'walk' this node yet

**Comment by KFlash** [comment]
@aladdin-add the tests tail because you are trying to walk the typesystem CST nodes. Do not walk them. It's not ready yet

'arr-pattern-with-type1.md'

File: src/forEachChild.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { NodeFlags, SyntaxKind } from './ast/syntax-node';
  *
  */
 
-export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKind): any {
+export function forEachChild(node: any, visitor: any): any {
```
Comment (KFlash) [inline src/forEachChild.ts:14]: what about this 'any' ?

File: src/forEachChild.ts [hunk 3]
```diff
@@ -107,8 +107,8 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
     case SyntaxKind.ArrowFunction:
       return (
         visitNode(node.arrowToken, node, visitor) ||
-        visitNode(node.typeParameters, node, visitor) ||
-        visitNode(node.parameters, node, visitor) ||
+        visitNodes(node.typeParameters, node, visitor) ||
+        visitNodes(node.parameters, node, visitor) ||
         visitNode(node.asyncToken, node, visitor) ||
```
Comment (KFlash) [inline src/forEachChild.ts:112]: can't walk a 'childLess' CST node. It doesn't have any children

File: src/forEachChild.ts [hunk 4]
```diff
@@ -107,8 +107,8 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
     case SyntaxKind.ArrowFunction:
       return (
         visitNode(node.arrowToken, node, visitor) ||
```
Comment (KFlash) [inline src/forEachChild.ts:109]: can't walk a 'childLess' arrowToken node CST node. It doesn't have any children. and 'parameters' is safe to walk

File: src/forEachChild.ts [hunk 5]
```diff
@@ -306,22 +306,22 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
...
         visitNode(node.contents, node, visitor) ||
-        visitNode(node.typeParameters, node, visitor) ||
+        visitNodes(node.typeParameters, node, visitor) ||
         visitNode(node.returnType, node, visitor)
       );
     case SyntaxKind.MethodDefinition:
       return (
-        visitNode(node.typeParameters, node, visitor) ||
-        visitNode(node.formalParameters, node, visitor) ||
+        visitNodes(node.typeParameters, node, visitor) ||
```
Comment (KFlash) [inline src/forEachChild.ts:316]: Should not walk this CST node *yet*

File: src/forEachChild.ts [hunk 6]
```diff
@@ -306,22 +306,22 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
...
+        visitNodes(node.formalParameters, node, visitor) ||
         visitNode(node.type, node, visitor) ||
         visitNode(node.contents, node, visitor)
       );
     case SyntaxKind.FunctionType:
       return (
-        visitNode(node.typeParameters, node, visitor) ||
-        visitNode(node.parameters, node, visitor) ||
+        visitNodes(node.typeParameters, node, visitor) ||
```
Comment (KFlash) [inline src/forEachChild.ts:323]: Should not walk this CST node *yet*

File: src/forEachChild.ts [hunk 7]
```diff
@@ -306,22 +306,22 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
...
+        visitNodes(node.parameters, node, visitor) ||
         visitNode(node.returnType, node, visitor)
```
Comment (KFlash) [inline src/forEachChild.ts:325]: Should not walk this CST node *yet*

File: src/forEachChild.ts [hunk 8]
```diff
@@ -345,18 +345,18 @@ export function forEachChild(node: any, visitor: (node: SyntaxKind) => SyntaxKin
 }
 
 function visitNode(
-  node: SyntaxKind,
+  node: any,
   parentNode: any,
-  visitor: (node: SyntaxKind, parentNode: SyntaxKind) => SyntaxKind
+  visitor: (node: any, parentNode: any) => SyntaxKind
```
Comment (KFlash) [inline src/forEachChild.ts:350]: What about 'any'?

**Review by KFlash — CHANGES_REQUESTED** [review]

File: src/forEachChild.ts (hunk 8 above)
Comment (aladdin-add) [reply src/forEachChild.ts:350]: the typings is incorrect, but I don't have to fix it atm.╮(￣▽￣)╭

File: test/runner/tob.ts [hunk 9]
```diff
@@ -48,7 +48,7 @@ export async function file2Tob(filename: string): Promise<Tob> {
   const cst = (tob.parserOptions.module ? parseModule : parseScript)(tob.input, tob.parserOptions, cb);
 
   // assert the cst is valid
-  visit(cst, tob.filename);
+  tob.parserOptions.allowTypes || visit(cst, tob.filename);
```
Comment (KFlash) [inline test/runner/tob.ts:51]: what's the purpose with this option? 'allowTypes' enables parsing type annotations like TS and Flow

Comment (aladdin-add) [reply test/runner/tob.ts:51]: let's focus on fixing JS-related issues, this is to skip walking type.

Comment (KFlash) [reply test/runner/tob.ts:51]: agree. types is still on my TODO

**Comment by KFlash** [comment]
@aladdin-add The tests are passing now :)

**Review by KFlash — APPROVED** [review]

**Comment by KFlash** [comment]
@aladdin-add Before you merge this, can you update the readme and add a section about the walker with an example on how to walk?

**Review by KFlash — CHANGES_REQUESTED** [review]
The walker isn't walking the CST nodes even they exist in the walker

not-walking

**Comment by aladdin-add** [comment]
what do you think merging this for now (the issue was already in the main branch)?

I can file an issue to track the rest of the TODOs.

**Comment by KFlash** [comment]
What is this progress bar? 

It's ok to merge it now. I checked and seems like the code coverage is the one to blame. It doesn't calculate the lines for this code even it's working

**Comment by KFlash** [comment]
@aladdin-add Look at the repo. What is this security issue?

**Comment by aladdin-add** [comment]
seems it was from `pretter-error`'s dev deps. I think we just remove it, as it was not used at all.

**Pull request merged**